File: src/main/java/net/miarma/mkernel/MKernel.java
package net.miarma.mkernel;

import java.io.File;
import java.nio.file.Files;
import java.util.logging.Logger;

import org.bukkit.plugin.java.JavaPlugin;

import dev.jorel.commandapi.CommandAPI;
import dev.jorel.commandapi.CommandAPIPaperConfig;
import net.miarma.mkernel.commands.CommandHandler;
import net.miarma.mkernel.common.minecraft.inventories.GlobalChest;
import net.miarma.mkernel.config.ConfigWrapper;
import net.miarma.mkernel.config.CustomConfigManager;
import net.miarma.mkernel.events.EventListener;
import net.miarma.mkernel.recipes.RecipeManager;
import net.miarma.mkernel.tasks.LocationTrackerTask;
import net.miarma.mkernel.util.FileUtil;

public class MKernel extends JavaPlugin {

    public static MKernel PLUGIN;
    public static final ConfigWrapper CONFIG = new ConfigWrapper();
    public static CustomConfigManager HOME_CONFIG;
    public static CustomConfigManager WORLD_BLOCKER_CONFIG;
    public static Logger LOGGER;

    @Override
    public void onLoad() {
        CommandAPI.onLoad(
    		new CommandAPIPaperConfig(this)
        		.verboseOutput(true)
        		.setNamespace("mkernel")
		);
    }
    
    @Override    
    public void onEnable() {
        // main plugin stuff
    	super.onEnable();
        PLUGIN = this;
        LOGGER = PLUGIN.getLogger();
        CONFIG.onEnable();
                
        // files handling
        HOME_CONFIG = new CustomConfigManager(PLUGIN, "homes.yml");
        WORLD_BLOCKER_CONFIG = new CustomConfigManager(PLUGIN, "blockedWorlds.yml");

        if(!Files.exists(PLUGIN.getDataFolder().toPath().resolve("inventories/"))) {
            File file = new File(PLUGIN.getDataFolder(), "inventories/");
            file.mkdirs();
        }

        if(!Files.exists(PLUGIN.getDataFolder().toPath().resolve("warps/"))) {
            File file = new File(PLUGIN.getDataFolder(), "warps/");
            file.mkdirs();
        }

        FileUtil.createLangs("lang.yml");
        
        // onEnable methods
        CommandAPI.onEnable();
        CommandHandler.onEnable();
        RecipeManager.onEnable();
        EventListener.onEnable();
        
        // final stage: location tracker and GC
        LocationTrackerTask.start();
        
        GlobalChest.loadConfig();
        GlobalChest.loadChest();
        
        PLUGIN.getLogger().info("I've been enabled! :)");
    }

    @Override
    public void onDisable() {
        GlobalChest.saveChest();
        CommandAPI.onDisable();
        PLUGIN.getLogger().info("I've been disabled! :(");
    }
}

File: src/main/kotlin/MKernel.kt
import com.google.inject.Guice
import com.google.inject.Injector
import dev.jorel.commandapi.CommandAPI
import dev.jorel.commandapi.CommandAPIPaperConfig
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import net.miarma.mkernel.command.CommandHandler
import net.miarma.mkernel.common.inject.MKernelModule
import net.miarma.mkernel.common.integration.HookLoader
import net.miarma.mkernel.common.integration.impl.WorldGuardHook
import net.miarma.mkernel.common.module.ModuleLoader
import net.miarma.mkernel.common.recipe.RecipeLoader
import net.miarma.mkernel.common.service.ServiceLoader
import net.miarma.mkernel.task.LocationTrackerTask
import net.miarma.mkernel.util.BukkitDispatcher
import org.bukkit.plugin.java.JavaPlugin
import java.util.logging.Logger
import kotlin.coroutines.CoroutineContext

class MKernel : JavaPlugin(), CoroutineScope {

    companion object {
        lateinit var PLUGIN: MKernel
            private set

        lateinit var LOGGER: Logger
            private set
    }

    private lateinit var job: Job
    override val coroutineContext: CoroutineContext
        get() = job + Dispatchers.Default

    lateinit var syncDispatcher: BukkitDispatcher
        private set

    private lateinit var injector: Injector

    override fun onLoad() {
        CommandAPI.onLoad(
            CommandAPIPaperConfig(this)
                .verboseOutput(false)
                .setNamespace("mkernel")
        )

        WorldGuardHook.registerFlag()
    }

    override fun onEnable() {
        PLUGIN = this
        LOGGER = logger

        job = Job()
        syncDispatcher = BukkitDispatcher(this)
        injector = Guice.createInjector(MKernelModule(this))

        injector.getInstance(ServiceLoader::class.java).loadAll()
        injector.getInstance(HookLoader::class.java).loadAll()
        injector.getInstance(ModuleLoader::class.java).loadAll()
        injector.getInstance(CommandHandler::class.java).registerCommands()
        injector.getInstance(RecipeLoader::class.java).loadAll()
        injector.getInstance(LocationTrackerTask::class.java).start()

        LOGGER.info("I've been enabled! :)")
    }

    override fun onDisable() {
        job.cancel()
        LOGGER.info("I've been disabled! :(")
    }

    fun launchAsync(block: suspend CoroutineScope.() -> Unit) {
        launch(Dispatchers.Default, block = block)
    }

    fun launchSync(block: suspend CoroutineScope.() -> Unit) {
        launch(syncDispatcher, block = block)
    }
}